File: GO/DFA_Toolkit/StatePartition.go
package dfatoolkit

// Block struct which represents a block within a partition.
type Block struct {
	Root    int        // Parent block of state.
	Size    int        // Size (score) of block.
	Link    int        // Index of next state within the block.
	Label   StateLabel // Label of block.
	Changed bool       // Whether block has been changed.
	Transitions []int  // Transition Table where each element corresponds to a transition for each symbol.
}

// StatePartition struct which represents a State Partition.
type StatePartition struct {
	Blocks               []Block // Slice of blocks.
	BlocksCount          int     // Number of blocks within partition.
	AcceptingBlocksCount int     // Number of accepting blocks within partition.
	RejectingBlocksCount int     // Number of rejecting blocks within partition.
	AlphabetSize int			 // Size of alphabet.
	StartingStateID int			 // The ID of the starting state.

	IsCopy             bool  // Whether state partition is a copy (for reverting merges).
	ChangedBlocks      []int // Slice of changed blocks.
	ChangedBlocksCount int   // Number of changed blocks within partition.
}

// NewStatePartition returns an initialized State Partition.
func NewStatePartition(referenceDFA DFA) StatePartition {
	// Initialize new State Partition struct with required values.
	statePartition := StatePartition{
		Blocks:          make([]Block, len(referenceDFA.States)),
		BlocksCount:     len(referenceDFA.States),
		AlphabetSize:    len(referenceDFA.Alphabet),
		StartingStateID: referenceDFA.StartingStateID,
	}

	// Set root and link as element, and size (score) as 1. Set block label
	// to state label and increment number of labelled states accordingly.
	// Transitions are then created using corresponding reference DFA.
	for i := 0; i < len(referenceDFA.States); i++ {
		statePartition.Blocks[i].Root = i
		statePartition.Blocks[i].Size = 1
		statePartition.Blocks[i].Link = i
		statePartition.Blocks[i].Label = referenceDFA.States[i].Label
		statePartition.Blocks[i].Changed = false
		if statePartition.Blocks[i].Label == ACCEPTING {
			statePartition.AcceptingBlocksCount++
		} else if statePartition.Blocks[i].Label == REJECTING {
			statePartition.RejectingBlocksCount++
		}

		// Initialize transitions.
		statePartition.Blocks[i].Transitions = make([]int, statePartition.AlphabetSize)
		for symbol := range referenceDFA.Alphabet{
			statePartition.Blocks[i].Transitions[symbol] = referenceDFA.States[i].Transitions[symbol]
		}
	}

	// Return initialized partition.
	return statePartition
}

// ChangedBlock updates the required fields to mark block as changed.
func (statePartition *StatePartition) ChangedBlock(blockID int) {
	// Check that state partition is a copy.
	if statePartition.IsCopy{
		// If block is not already changed.
		if !statePartition.Blocks[blockID].Changed {
			// Update changed slice to include changed block ID.
			statePartition.ChangedBlocks[statePartition.ChangedBlocksCount] = blockID
			// Increment the changed blocks counter.
			statePartition.ChangedBlocksCount++
			// Set changed flag within block to true.
			statePartition.Blocks[blockID].Changed = true
		}
	}
}

// Union connects two blocks by comparing their respective
// size (score) values to keep the tree flat.
func (statePartition *StatePartition) Union(blockID1 int, blockID2 int){
	// Mark both blocks as changed so merge can be undone.
	statePartition.ChangedBlock(blockID1)
	statePartition.ChangedBlock(blockID2)

	// Decrement blocks count.
	statePartition.BlocksCount--

	// If size of parent node is smaller than size of child node, switch
	// parent and child nodes.
	if statePartition.Blocks[blockID1].Size < statePartition.Blocks[blockID2].Size{
		blockID1, blockID2 = blockID2, blockID1
	}

	// Set pointer of block 1 to parent and pointer of block 2 to child.
	parent, child := &statePartition.Blocks[blockID1], &statePartition.Blocks[blockID2]

	// Link nodes by assigning the link of parent to link of child and vice versa.
	parent.Link, child.Link = child.Link, parent.Link

	// Set root of child node to parent node.
	child.Root = blockID1
	// Increment size (score) of parent node by size of child node.
	parent.Size += child.Size

	// If label of parent is unknown and label of child is
	// not unknown, set label of parent to label of child.
	if parent.Label == UNKNOWN && child.Label != UNKNOWN {
		parent.Label = child.Label
	} else if parent.Label == ACCEPTING && child.Label == ACCEPTING {
		// Else, if both blocks are accepting, decrement accepting blocks count.
		statePartition.AcceptingBlocksCount--
	} else if parent.Label == REJECTING && child.Label == REJECTING {
		// Else, if both blocks are rejecting, decrement rejecting blocks count.
		statePartition.RejectingBlocksCount--
	}

	// Update transitions.
	for i := 0; i < statePartition.AlphabetSize; i++ {
		// If transition of parent does not exist,
		// set transition of child and set child
		// transition to -1 (remove transition).
		if parent.Transitions[i] == -1 {
			parent.Transitions[i] = child.Transitions[i]
			child.Transitions[i] = -1
		}
	}
}

// Find traverses each root element while compressing the
// levels to find the root element of the stateID.
func (statePartition *StatePartition) Find(stateID int) int {
	// Traverse each root block until state is reached.
	for statePartition.Blocks[stateID].Root != stateID {
		// Compress if necessary.
		if statePartition.Blocks[stateID].Root != statePartition.Blocks[statePartition.Blocks[stateID].Root].Root{
			// If compression is required, mark state as
			// changed and set root to root of parent.
			statePartition.ChangedBlock(stateID)
			statePartition.Blocks[stateID].Root = statePartition.Blocks[statePartition.Blocks[stateID].Root].Root
		}
		stateID = statePartition.Blocks[stateID].Root
	}

	return stateID
}

// ReturnSet returns the state IDs within given block.
func (statePartition StatePartition) ReturnSet(blockID int) []int {
	// Slice of state IDs.
	blockElements := []int{blockID}
	// Set root to block ID.
	root := blockID

	// Iterate until link of current block ID is
	// not equal to the root block.
	for statePartition.Blocks[blockID].Link != root {
		// Set block ID to link of current block.
		blockID = statePartition.Blocks[blockID].Link
		// Add block ID to block elements slice.
		blockElements = append(blockElements, blockID)

		// Panic if length of elements slice is bigger than number of blocks.
		//if len(blockElements) > len(statePartition.Blocks) {
		//	panic("Error in state linking.")
		//}
	}

	// Return state IDs within block.
	return blockElements
}

// WithinSameBlock checks whether two states are within the same block.
func (statePartition *StatePartition) WithinSameBlock(stateID1 int, stateID2 int) bool {
	return statePartition.Find(stateID1) == statePartition.Find(stateID2)
}

// ToStatePartition converts a DFA to a State Partition.
func (dfa DFA) ToStatePartition() StatePartition {
	// Return NewStatePartition.
	return NewStatePartition(dfa)
}

// ToDFA converts a State Partition to a DFA. Returns true and
// the corresponding DFA if state partition is valid. Else,
// false and an empty DFA are returned.
func (statePartition StatePartition) ToDFA() DFA {
	// Map to store corresponding new state for
	// each root block within state partition.
	blockToStateMap := map[int]int{}

	// Initialize resultant DFA to be returned by function.
	resultantDFA := NewDFA()

	// Get root blocks within state partition.
	rootBlocks := statePartition.RootBlocks()

	// Create a new state within DFA for each root block and
	// set state label to block label.
	for _, stateID := range rootBlocks {
		blockToStateMap[stateID] = resultantDFA.AddState(statePartition.Blocks[stateID].Label)
	}

	// Create alphabet within DFA.
	for symbol := 0; symbol < statePartition.AlphabetSize; symbol++ {
		resultantDFA.AddSymbol()
	}

	// Update transitions using transitions within blocks and block to state map.
	for _, stateID := range rootBlocks {
		for symbol := 0; symbol < statePartition.AlphabetSize; symbol++ {
			if resultantState := statePartition.Blocks[stateID].Transitions[symbol]; resultantState > -1{
				resultantDFA.States[blockToStateMap[stateID]].Transitions[symbol] = blockToStateMap[statePartition.Find(resultantState)]
			}
		}
	}

	// Set starting state using block to state map.
	resultantDFA.StartingStateID = blockToStateMap[statePartition.Find(statePartition.StartingStateID)]

	// Return populated resultant DFA.
	return resultantDFA
}

// MergeStates recursively merges states to merge state1 and state2. Returns false if merge
// results in a non-deterministic automaton. Returns true if merge was successful.
func (statePartition *StatePartition) MergeStates(state1 int, state2 int) bool {
	// If parent blocks (root) are the same as state ID, skip finding the root.
	// Else, find the parent block (root) using Find function.
	if statePartition.Blocks[state1].Root != state1 {
		state1 = statePartition.Find(state1)
	}
	if statePartition.Blocks[state2].Root != state2 {
		state2 = statePartition.Find(state2)
	}

	// Return true if states are already in the same block
	// since merge is not required.
	if state1 == state2 {
		return true
	}

	// Get pointer of both blocks.
	block1, block2 := &statePartition.Blocks[state1],  &statePartition.Blocks[state2]

	// If labels are contradicting, return false since this results
	// in a non-deterministic automaton so merge cannot be done.
	if (block1.Label == ACCEPTING && block2.Label == REJECTING) || (block1.Label == REJECTING && block2.Label == ACCEPTING) {
		return false
	}

	// Merge states within state partition.
	statePartition.Union(state1, state2)

	// Iterate over alphabet.
	for i := 0; i < statePartition.AlphabetSize; i++ {
		// If either block1 or block2 do not have a transition, continue
		// since no merge is required.
		if block1.Transitions[i] == -1 || block2.Transitions[i] == -1{
			continue
		}
		// Else, merge resultant blocks.
		if !statePartition.MergeStates(block1.Transitions[i], block2.Transitions[i]) {
			return false
		}
	}

	// Return true if this is reached (deterministic).
	return true
}

// Copy copies the state partition.
func (statePartition StatePartition) Copy() StatePartition {
	// Initialize new StatePartition struct using state partition.
	copiedStatePartition := StatePartition{
		Blocks:               make([]Block, len(statePartition.Blocks)),
		BlocksCount:          statePartition.BlocksCount,
		AcceptingBlocksCount: statePartition.AcceptingBlocksCount,
		RejectingBlocksCount: statePartition.RejectingBlocksCount,
		AlphabetSize:         statePartition.AlphabetSize,
		StartingStateID:      statePartition.StartingStateID,
		IsCopy:               true,
		ChangedBlocks:        make([]int, len(statePartition.Blocks)),
		ChangedBlocksCount:   0,
	}

	// Iterate over each block within blocks slice.
	for blockID := range statePartition.Blocks{
		// Get block pointer from copied state partition.
		block := &copiedStatePartition.Blocks[blockID]
		// Get block pointer from original state partition.
		originalBlock := &statePartition.Blocks[blockID]

		// Update root, size, link, and label.
		block.Root = originalBlock.Root
		block.Size = originalBlock.Size
		block.Link = originalBlock.Link
		block.Label = originalBlock.Label

		// Initialize and copy transitions.
		block.Transitions = make([]int, statePartition.AlphabetSize)
		copy(block.Transitions, originalBlock.Transitions)
	}

	// Return copied state partition.
	return copiedStatePartition
}

// RollbackChanges reverts any changes made within state partition given the original state partition.
func (statePartition *StatePartition) RollbackChanges(originalStatePartition StatePartition) {
	// If the state partition is a copy, copy values of changed blocks from original
	// state partition. Else, do nothing.
	if statePartition.IsCopy {
		// Set blocks count values to the original values.
		statePartition.BlocksCount = originalStatePartition.BlocksCount
		statePartition.AcceptingBlocksCount = originalStatePartition.AcceptingBlocksCount
		statePartition.RejectingBlocksCount = originalStatePartition.RejectingBlocksCount

		// Iterate over each altered block.
		for _, blockID := range statePartition.ChangedBlocks[:statePartition.ChangedBlocksCount] {
			// Get block pointer from copied state partition.
			block := &statePartition.Blocks[blockID]
			// Get block pointer from original state partition.
			originalBlock := &originalStatePartition.Blocks[blockID]

			// Update root, size, link, and label.
			block.Root = originalBlock.Root
			block.Size = originalBlock.Size
			block.Link = originalBlock.Link
			block.Label = originalBlock.Label
			block.Changed = false

			// Copy transitions.
			copy(block.Transitions, originalBlock.Transitions)
		}

		// Set the changed blocks count to 0.
		statePartition.ChangedBlocksCount = 0
	}
}

// NumberOfLabelledBlocks returns the number of labelled blocks (states) within state partition.
func (statePartition StatePartition) NumberOfLabelledBlocks() int {
	// Return the sum of the accepting and rejecting blocks count.
	return statePartition.AcceptingBlocksCount + statePartition.RejectingBlocksCount
}

// RootBlocks returns the IDs of root blocks as a slice of integers.
func (statePartition StatePartition) RootBlocks() []int {
	// Initialize slice using blocks count value.
	rootBlocks := make([]int, statePartition.BlocksCount)
	// Index (count) of root blocks.
	index := 0

	// Iterate over each block within partition.
	for blockID := range statePartition.Blocks {
		// Check if root of current block is equal to the block ID
		if statePartition.Blocks[blockID].Root == blockID {
			// Add to rootBlocks slice using index.
			rootBlocks[index] = blockID
			// Increment index.
			index++
			// If index is equal to blocks count,
			// break since all root blocks have
			// been found.
			if index == statePartition.BlocksCount {
				break
			}
		}
	}

	// Return populated slice.
	return rootBlocks
}

// OrderedBlocks returns the IDs of root blocks in order as a slice of integers.
func (statePartition StatePartition) OrderedBlocks() []int{
	orderComputed := make([]bool, len(statePartition.Blocks))
	orderedBlocks := make([]int, statePartition.BlocksCount)
	index := 0

	// Create a FIFO queue with starting state.
	queue := []int{statePartition.Find(statePartition.Find(statePartition.StartingStateID))}

	// Loop until queue is empty.
	for len(queue) > 0 {
		// Remove and store first state in queue.
		blockID := queue[0]
		queue = queue[1:]

		if orderComputed[blockID]{
			continue
		}

		// Set the order of the current state.
		orderedBlocks[index] = blockID
		orderComputed[blockID] = true
		// Increment current state order.
		index++

		// Iterate over each symbol (alphabet) within DFA.
		for symbol := 0; symbol < statePartition.AlphabetSize; symbol++ {
			// If transition from current state using current symbol is valid and is not a loop to the current state.
			if childStateID := statePartition.Blocks[blockID].Transitions[symbol]; childStateID != -1 {
				// If depth for child state has been computed, skip state.
				if childBlockID := statePartition.Find(childStateID); childBlockID != blockID {
					// Add child state to queue.
					queue = append(queue, childBlockID)
				}
			}
		}
	}

	return orderedBlocks
}

File: Rust/dfa_toolkit/src/state_partition.rs
use crate::dfa::DFA;
use crate::dfa_state::{StateLabel, ACCEPTING, REJECTING, UNKNOWN};
use std::borrow::{Borrow, BorrowMut};

pub struct Block {
    pub root: i32,
    pub size: i32,
    pub link: i32,
    pub label: StateLabel,
    pub changed: bool,
    pub transitions: Vec<i32>,
}

pub struct StatePartition {
    pub blocks: Vec<Block>,
    pub blocks_count: i32,
    pub accepting_blocks_count: i32,
    pub rejecting_blocks_count: i32,
    pub alphabet_size: usize,
    pub starting_state_id: i32,

    pub is_copy: bool,
    pub changed_blocks: Vec<i32>,
    pub changed_blocks_count: i32,
}

impl StatePartition {
    pub fn new(dfa: &DFA) -> StatePartition {
        let mut state_partition = StatePartition {
            blocks: Vec::with_capacity(dfa.states.len()),
            blocks_count: dfa.states.len() as i32,
            accepting_blocks_count: 0,
            rejecting_blocks_count: 0,
            alphabet_size: dfa.symbols_count,
            starting_state_id: dfa.starting_state_id,
            is_copy: false,
            changed_blocks: vec![],
            changed_blocks_count: 0,
        };

        for i in 0..dfa.states.len() {
            let mut block = Block {
                root: i as i32,
                size: 1,
                link: i as i32,
                label: dfa.states[i].label,
                changed: false,
                transitions: Vec::with_capacity(dfa.symbols_count as usize),
            };

            if block.label == ACCEPTING {
                state_partition.accepting_blocks_count += 1;
            } else if block.label == REJECTING {
                state_partition.rejecting_blocks_count += 1;
            }

            for j in 0..dfa.symbols_count {
                block
                    .transitions
                    .push(dfa.states[i].transitions[j as usize]);
            }

            state_partition.blocks.push(block);
        }

        return state_partition;
    }

    pub fn copy(&self) -> StatePartition {
        let mut copied_state_partition = StatePartition {
            blocks: Vec::with_capacity(self.blocks.len()),
            blocks_count: self.blocks_count,
            accepting_blocks_count: self.accepting_blocks_count,
            rejecting_blocks_count: self.rejecting_blocks_count,
            alphabet_size: self.alphabet_size,
            starting_state_id: self.starting_state_id,
            is_copy: true,
            changed_blocks: vec![0; self.blocks.len()],
            changed_blocks_count: 0,
        };

        for i in 0..self.blocks.len() {
            let block_pointer = self.blocks[i].borrow();
            copied_state_partition.blocks.push(Block {
                root: block_pointer.root,
                size: block_pointer.size,
                link: block_pointer.link,
                label: block_pointer.label,
                changed: false,
                transitions: block_pointer.transitions.to_vec(),
            });
        }

        return copied_state_partition;
    }

    pub fn changed_block(&mut self, block_id: i32) {
        // Check that state partition is a copy and that block is not modified.
        if self.is_copy && !self.blocks[block_id as usize].changed {
            // Update changed vector to include changed block ID.
            self.changed_blocks[self.changed_blocks_count as usize] = block_id;
            // Increment the changed blocks counter.
            self.changed_blocks_count += 1;
            // Set changed flag within block to true.
            self.blocks[block_id as usize].changed = true;
        }
    }

    pub fn union(&mut self, mut block_id_1: i32, mut block_id_2: i32) {
        // If state partition is a copy, call ChangedBlock for
        // both blocks so merge can be undone if necessary.
        self.changed_block(block_id_1);
        self.changed_block(block_id_2);

        // Decrement blocks count.
        self.blocks_count -= 1;

        // If size of parent node is smaller than size of child node, switch
        // parent and child nodes.
        if self.blocks[block_id_1 as usize].size < self.blocks[block_id_2 as usize].size {
            let temp_value = block_id_1;
            block_id_1 = block_id_2;
            block_id_2 = temp_value;
        }

        // Link nodes by assigning the link of parent to link of child and vice versa.
        let temp_value = self.blocks[block_id_1 as usize].link;
        self.blocks[block_id_1 as usize].link = self.blocks[block_id_2 as usize].link;
        self.blocks[block_id_2 as usize].link = temp_value;

        // Set root of child node to parent node.
        self.blocks[block_id_2 as usize].root = block_id_1;
        // Increment size (score) of parent node by size of child node.
        self.blocks[block_id_1 as usize].size += self.blocks[block_id_2 as usize].size;

        // If label of parent is unknown and label of child is
        // not unknown, set label of parent to label of child.
        if self.blocks[block_id_1 as usize].label == UNKNOWN
            && self.blocks[block_id_2 as usize].label != UNKNOWN
        {
            self.blocks[block_id_1 as usize].label = self.blocks[block_id_2 as usize].label;
        } else if self.blocks[block_id_1 as usize].label == ACCEPTING
            && self.blocks[block_id_2 as usize].label == ACCEPTING
        {
            // Else, if both blocks are accepting, decrement accepting blocks count.
            self.accepting_blocks_count -= 1;
        } else if self.blocks[block_id_1 as usize].label == REJECTING
            && self.blocks[block_id_2 as usize].label == REJECTING
        {
            // Else, if both blocks are rejecting, decrement rejecting blocks count.
            self.rejecting_blocks_count -= 1;
        }

        // Update transitions.
        for i in 0..self.alphabet_size {
            // If transition of parent does not exist,
            // set transition of child and set child
            // transition to -1 (remove transition).
            if self.blocks[block_id_1 as usize].transitions[i] == -1 {
                self.blocks[block_id_1 as usize].transitions[i] =
                    self.blocks[block_id_2 as usize].transitions[i];
                self.blocks[block_id_2 as usize].transitions[i] = -1;
            }
        }
    }

    pub fn find(&mut self, state: i32) -> i32 {
        let mut state_id = state;

        // Traverse each root block until state is reached.
        while self.blocks[state_id as usize].root != state_id {
            // Compress if necessary.
            if self.blocks[state_id as usize].root
                != self.blocks[self.blocks[state_id as usize].root as usize].root
            {
                // If compression is required, mark state as
                // changed and set root to root of parent.
                self.changed_block(state_id);
                self.blocks[state_id as usize].root =
                    self.blocks[self.blocks[state_id as usize].root as usize].root;
            }
            state_id = self.blocks[state_id as usize].root;
        }

        return state_id;
    }

    // pub fn return_set(&self, block: i32) -> Vec<i32>{
    //     let mut block_id = block;
    //
    //     // Hash set of state IDs and add root element to set.
    //     let mut block_elements: Vec<i32> = vec![block_id];
    //
    //     // Add root element to set.
    //     //block_elements.push(block_id);
    //     // Set root to block ID.
    //
    //     let root = block_id;
    //
    //     // Iterate until link of current block ID is
    //     // not equal to the root block.
    //     while self.blocks[block_id as usize].link != root{
    //         // Set block ID to link of current block.
    //         block_id = self.blocks[block_id as usize].link;
    //         // Add block ID to block elements set.
    //         block_elements.push(block_id);
    //     }
    //
    //     // Return state IDs within set.
    //     return block_elements;
    // }

    pub fn merge_states(&mut self, state1: i32, state2: i32) -> bool {
        let mut state1_id = state1;
        let mut state2_id = state2;

        // If parent blocks (root) are the same as state ID, skip finding the root.
        // Else, find the parent block (root) using Find function.
        if self.blocks[state1_id as usize].root != state1_id {
            state1_id = self.find(state1_id);
        }
        if self.blocks[state2_id as usize].root != state2_id {
            state2_id = self.find(state2_id);
        }

        // Return true if states are already in the same block
        // since merge is not required.
        if state1_id == state2_id {
            return true;
        }

        // Get pointer of both blocks.
        //let (mut block1, mut block2) = (&self.blocks[state1_id as usize], &self.blocks[state2_id as usize]);

        // If labels are contradicting, return false since this results
        // in a non-deterministic automaton so merge cannot be done.
        if (self.blocks[state1_id as usize].label == ACCEPTING
            && self.blocks[state2_id as usize].label == REJECTING)
            || (self.blocks[state1_id as usize].label == REJECTING
                && self.blocks[state2_id as usize].label == ACCEPTING)
        {
            return false;
        }

        // Merge states within state partition.
        self.union(state1_id, state2_id);

        // Iterate over alphabet.
        for i in 0..self.alphabet_size {
            // If either block1 or block2 do not have a transition, continue
            // since no merge is required.
            if self.blocks[state1_id as usize].transitions[i] == -1
                || self.blocks[state2_id as usize].transitions[i] == -1
            {
                continue;
            }
            // Else, merge resultant blocks.
            if !self.merge_states(
                self.blocks[state1_id as usize].transitions[i],
                self.blocks[state2_id as usize].transitions[i],
            ) {
                return false;
            }
        }

        // Return true if this is reached (deterministic).
        return true;
    }

    pub fn rollback_changes(&mut self, original_state_partition: &StatePartition) {
        // If the state partition is a copy, copy values of changed blocks from original
        // state partition. Else, do nothing.
        if self.is_copy {
            // Set blocks count values to the original values.
            self.blocks_count = original_state_partition.blocks_count;
            self.accepting_blocks_count = original_state_partition.accepting_blocks_count;
            self.rejecting_blocks_count = original_state_partition.rejecting_blocks_count;

            // Iterate over each altered block (state).
            for i in 0..self.changed_blocks_count {
                let block_id = self.changed_blocks[i as usize];
                // Get block pointer from copied state partition.
                let mut block = self.blocks[block_id as usize].borrow_mut();
                // Get block pointer from original state partition.
                let original_block = original_state_partition.blocks[block_id as usize].borrow();

                // Update root, size, link, and label.
                block.root = original_block.root;
                block.size = original_block.size;
                block.link = original_block.link;
                block.label = original_block.label;
                block.changed = false;
                block.transitions = original_block.transitions.to_vec();
            }

            // Empty the changed blocks vector.
            self.changed_blocks_count = 0;
        }
    }
}

impl DFA {
    pub fn to_state_partition(&self) -> StatePartition {
        return StatePartition::new(self);
    }
}
